lib/Sema/CSSimplify.cpp

```diff
@@ -3939,9 +3940,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
     break;
   }
   
+  // FIXME: Allow the type to be upcast if the type system has a concrete
+  // KeyPath type assigned to the expression already.
+  if (keyPathBGT) {
+    if (keyPathBGT->getDecl() == getASTContext().getKeyPathDecl())
+      kpDecl = getASTContext().getKeyPathDecl();
+    else if (keyPathBGT->getDecl() == getASTContext().getWritableKeyPathDecl()
+             && capability >= Writable)
+      kpDecl = getASTContext().getWritableKeyPathDecl();
+  }
```

Comment:
If I remove this, then these tests fail in `test/expr/unary/keypath/keypath.swift`:

https://github.com/jckarter/swift/blob/35527847b44d37aedb27b95be3080f89770c747d/test/expr/unary/keypath/keypath.swift#L94
https://github.com/apple/swift/pull/10773/files#diff-cb09d088767b84ab589238604bee89edR349

because it looks like, by the time the constraint solver tries simplifying the key path constraint in these places, it's already bound the type variable for the KeyPathExpr's type to the contextual type.